### DISCUSSION TITLE
Fisher Discriminant Analysis

### DISCUSSION THREAD

**Discussion by gianluvilla** [StevenGolovkine/FDApy#24] [Ideas]
Hi,
First of all, I want to thank you for your work on this package, it’s been working great so far!

I am currently working with a set of images (dosimetric maps) related to patients who may experience side effects from radiotherapy. So far, I have applied FPCA on a 2-dimensional domain to see if the FPCA scores correlate with the presence of side effects.

I was wondering if there is a way to apply something like Fisher’s Discriminant Analysis to these two clusters of data (patients with and without side effects), in order to directly extract the features from the images that best separate the two classes.

Thank you so much, and have a great day!

**Comment by StevenGolovkine** [comment]
Hi @gianluvilla,

Thanks for your comment and sorry for the late answer.
To be honest, I don't know the literature very well on this topic, but you may find these papers interesting:

- Common FPCA: https://www.sciencedirect.com/science/article/abs/pii/S0167945710001867
- LDA: https://pmc.ncbi.nlm.nih.gov/articles/PMC9042036/

So, for now, nothing is implemented in the package in this regard.
However, I was planning to extend the Common FPCA idea to multidimensional and multivariate functional data.
If you would be interested in collaborating on this topic, please email me.

Have a nice day!

**Comment by gianluvilla** [comment]
Hi, thank you for your response
I will definitely check both links; Functional LDA, in particular, seems to be exactly what I was looking for, as it should be equivalent to Fisher’s Discriminant Analysis under certain assumptions.

I was wondering if there are any existing implementations available in Python, even though I understand it’s quite a niche topic.

Thank you for your proposal, but unfortunately, I’m just a student in Mathematical Engineering, so I don’t feel competent enough to contribute to the implementation of such a package at this time.

Thank you again for your help, and have a great day!

**Comment by gianluvilla** [comment]
Hi @StevenGolovkine,
I just wanted to let you know that the project I've been working on using your package will be presented next week at the ESTRO conference (European SocieTy for Radiotherapy and Oncology), and will likely be published as a research paper. I'm really proud of this result, and I want to thank you again for developing and sharing this package.

I did notice something during the image reconstruction phase (.inverse_transform(scores)). The algorithm seems to overestimate the higher values. For example, if the original data ranges from 0 to 80, regions with true values around 70 tend to be reconstructed as 75–80. This could be an issue for my specific application, since the higher values are the most important and need to be reconstructed as accurately as possible.

Is there any trick or adjustment I could apply to make the algorithm more sensitive to accurately reconstructing these higher-value regions?

Thanks again, have a good day. 
Gianluca

**Reply by StevenGolovkine** [reply]
Hi @gianluvilla, 

Congrats for the presentation. Send me the paper when it is ready!

Concerning your issue, it is hard to say without seeing the data. Do you have localized peak in your surface? Does the surface is relatively smooth? Maybe, the number of components you retain is not important enough to retrieve these high values. Maybe, your sampling grid is not precise enough to get these values. Maybe you should try to normalize the data beforehand. 
There are a lot of different possibilities!

Have a nice day

**Reply by gianluvilla** [reply]
Hi, thank you for your response. I've been thinkig on normalizing the data, but i wanted to give more "weight" to the higher-values regions, and i'm worried that normalizing data before fitting the fpca would just amplify this error. 

Here is the visualization of the isolines for original (smoothed) image and the reconstructed one (without applying normalization, using 10 principal components). As you can see, in the reconstructed image we have values > 80 in the central region, but the original map doesn't have any pixel with value > 80. 
![Immagine 2025-05-05 135238](https://github.com/user-attachments/assets/5e2f8db4-cbf1-4644-9de4-4df3ce6136c1)

Thank you for your time, have a nice day.

Gianluca

**Reply by StevenGolovkine** [reply]
You're probably right, the normalization won't solve your problem.  From what I see, you can try to increase the number of components to retrieve, it might help. 
However, the reconstruction will never be perfect because of the different estimation steps and it is usually difficult to estimate the regions equal to 0 because the eigenfunctions have to cancel each other in these areas.

(PS: You have a typo in your caption.)

**Reply by gianluvilla** [reply]
The thing is, I would expect the reconstruction to underestimate these higher-valued regions, since they’re less “conformal.” Moreover, since the component scores decrease in magnitude as the component number increases, they contribute less and less to the reconstruction. As a result, this issue often isn’t resolved by adding more components.
As you said, the problem might be caused by other regions, like the ones that have really low variance (because here, the deviation from the mean will be zero and therefore harder to reconstruct). I’ll try excluding these regions from the analysis to see if the problem is mitigated.

Thanks again, have a good day!
(P.S.: Thanks, I hadn’t realized that)
